plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -130,6 +229,127 @@ class bp_connection_manager {
              established_client_connection(new_connection) && num_established_clients() > self()->connections.get_max_client_count();
    }
 
+   // thread safe
+   // removes invalid entries from msg
+   bool validate_gossip_bp_peers_message( gossip_bp_peers_message& msg ) {
+      if (msg.peers.empty())
+         return false;
+      if (msg.peers.size() != 1 || !msg.peers[0].server_address.empty()) { // initial case, no server_addresses to validate
+         auto valid_address = [](const std::string& addr) -> bool {
+            const auto& [host, port, type] = net_utils::split_host_port_type(addr);
+            return !host.empty() && !port.empty();
+         };
+         std::map<eosio::name, size_t> producers;
+         const gossip_bp_peers_message::bp_peer* prev = nullptr;
+         for (const auto& peer : msg.peers) {
+            if (peer.producer_name.empty())
+               return false;
+            if (!valid_address(peer.server_address))
+               return false;
+            if (prev != nullptr) {
+               if (prev->producer_name == peer.producer_name) {
+                  if (prev->server_address == peer.server_address)
+                     return false; // duplicate entries not allowed
+               } else if (prev->producer_name > peer.producer_name) {
+                  return false; // required to be sorted
+               }
+            }
+            if (++producers[peer.producer_name] > 4)
+               return false; // only 4 entries per producer allowed
+            prev = &peer;
+         }
+      }
+
+      controller& cc = self()->chain_plug->chain();
+
+      fc::lock_guard g(gossip_bps.mtx);
+      auto& sig_idx = gossip_bps.index.get<by_sig>();
+      for (auto i = msg.peers.begin(); i != msg.peers.end();) {
+         const auto& peer = *i;
+         try {
```

Comment:
Can the `try/catch` block be moved into the if below (line 270)?

Also instead of all these `erase(); continue;`, I think it would be somewhat cleaner to have a lambda:
```
auto is_peer_msg_valid = [&](const bp_peer& peer) -> bool { ...};
```

so the loop would become:

```
for (auto i = msg.peers.begin(); i != msg.peers.end();) {
   if (!is_peer_msg_valid(*i) 
      i = msg.peers.erase(i);
   else
      ++i;
}
```